### PR TITLE
Shrink the prelude

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Currently, the prelude contains basically everything, which can cause name clashes when building third party crates. This PR shrinks the default prelude and adds dedicated, additional preludes for 2d and 3d.

### CODE REVIEW

File: crates/bevy_internal/src/prelude.rs [hunk 1]
```diff
@@ -1,44 +1,8 @@
...
 pub use crate::{
-    app::prelude::*, asset::prelude::*, core::prelude::*, ecs::prelude::*, input::prelude::*,
-    log::prelude::*, math::prelude::*, reflect::prelude::*, scene::prelude::*,
-    transform::prelude::*, window::prelude::*, DefaultPlugins, MinimalPlugins,
+    app::prelude::*, asset::prelude::*, core::prelude::*, ecs::prelude::*, log::prelude::*,
+    math::prelude::*, render::prelude::*, transform::prelude::*, window::prelude::*,
+    DefaultPlugins, MinimalPlugins,
 };
 
 pub use bevy_derive::bevy_main;
```
Comment (bjorn3) [inline crates/bevy_internal/src/prelude.rs:8]: Maybe inline this into `lib.rs` too?

File: crates/bevy_internal/src/lib.rs [hunk 2]
```diff
@@ -141,3 +141,17 @@ pub mod dynamic_plugin {
...
+    pub use crate::prelude::*;
+    #[cfg(feature = "bevy_sprite")]
+    pub use crate::sprite::prelude::*;
+}
+
+pub mod prelude3d {
+    #[cfg(feature = "bevy_pbr")]
+    pub use crate::pbr::prelude::*;
+    pub use crate::prelude::*;
+}
```
Comment (bjorn3) [inline crates/bevy_internal/src/lib.rs:157]:
```suggestion
pub mod prelude2d {
    pub use crate::prelude::*;

    #[cfg(feature = "bevy_core_pipeline")]
    pub use crate::core_pipeline::ClearColor;
    #[cfg(feature = "bevy_sprite")]
    pub use crate::sprite::prelude::*;
}

pub mod prelude3d {
    pub use crate::prelude::*;

    #[cfg(feature = "bevy_pbr")]
    pub use crate::pbr::prelude::*;
}
```
This makes it easier to see what `#[cfg]` applies to every `use` at a glance.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_internal/src/lib.rs [hunk 3]
```diff
@@ -141,3 +138,31 @@ pub mod dynamic_plugin {
...
+    pub use crate::{
+        app::prelude::*, asset::prelude::*, core::prelude::*, ecs::prelude::*, log::prelude::*,
+        math::prelude::*, render::prelude::*, transform::prelude::*, window::prelude::*,
+        DefaultPlugins, MinimalPlugins,
+    };
+
+    pub use bevy_derive::bevy_main;
+}
+
+pub mod prelude2d {
```
Comment (aloucks) [inline crates/bevy_internal/src/lib.rs:154]: Other modules that end with `2d` and `3d` have an underscore prefix. These might be better as `prelude_2d` and `prelude_3d` to remain consistent with other parts of the API.

https://github.com/bevyengine/bevy/blob/adb3ad399cf092110b2d69f7dc189abc0ebc9f80/crates/bevy_core_pipeline/src/lib.rs#L3-L4

**Comment by mockersf** [comment]
> additional preludes for 2d and 3d.

If this two need to be different, we should document more how people can toggle features to select what they want rather than having more preludes.

**Comment by alice-i-cecile** [comment]
>> additional preludes for 2d and 3d.
>
> If these two need to be different, we should document more how people can toggle features to select what they want rather than having more preludes.


I don't think that features are the right tool here. When making 2D-centric games, you'll often want to splash in occasional 3D functionality or vice versa. Completely disabling the other mode with features forbids this, but preludes cover the 90% case well.

**Pull request closed without merging**